accepted/rfc0004-logs.md

```diff
@@ -0,0 +1,203 @@
+- change `in_registry_cold` and `in_registry_cross` to create an initial task
+- change `join` to log the raw events
+- provide a `subgraph` function
+- provide the conversion function from raw events to tasks
+- provide a `save_logs` function
+
+All graph and svg related functions are not here to keep the changes small.
+
+All modifications are in `rayon-core` only.
+All modifications are in included in `cfg` which means you'll fall back to the standard rayon if you do not enable the `logs` feature.
+
+All `scope` and `spawn` are not implemented yet. It shouldn't be long but I'd rather like some feedback first.
+
+# Implementation notes
+
+Files coming from `rayon-logs` are in the `logs` subdirectory.
+
+- `subgraph.rs` contains the subgraph functions
+- `raw_events.rs` the definitions of the raw events
+- `runlog.rs` the tasks logs definitions and conversion function
+- `storage.rs` the structure used for storing raw events on each thread
+
+Then the modified files in `rayon-core` are currently:
+
+- `join/mod.rs` for the logged join implementation
+- `registry/mod.rs` for modifications to the `Registry` structure, the `in_registry` functions and the global `save_logs` function.
+
+On the other side `rayon-logs` has been adapted to allow disconnected graphs.
+
+## Implementing `Storage`
+
+The `Storage` structure is a linked list of vectors (capacitated).
+The idea is that most of the time you just push the event with a very little cost.
+However should you log a lot of events you will have another block re-allocated for you, still in O(1) complexity (well, plus the cost of
+the memory allocation).
+
+It contains an `UnsafeCell` since threads write events to it but someone is going to read the logs and save them.
+Since the vectors are never re-allocated reading a block is always safe. We could have problems
+with lists pointers but I use an atomic counter for the number of blocks which is incremented only after a pointer is changed.
+So, this should hopefully also be safe.
+
+Currently logs are never erased. Calling `save_logs` several times will re-save some tasks several times.
+
+## Implementing `join`
+
+We changed the join function in the following way:
+
+- we use the `in_worker` function to get access to the current thread
+- we go up to the registry and use the thread index to push events in the corresponding storage
+- `op_a` and `op_b` are hijacked to log a new task start at their start and a new task end at their end
+- the current task is ended just before the join and restarted just after
+- we tag the links in the graph between the tasks
+
+# Unresolved questions
+
+## Crates Dependencies
+
+Currently I depend on the following crates. Should I remove these dependencies ?
+- time (easy to remove)
+- itertools (harder to remove but ok)
+- serde (also easy to remove since I output logs in json)
+
+## Svg
+
+All graph algorithms and svg display are not included.

```

Comment:
I think we will want to create an external crate that is able to do the postprocessing and imagine manipulation